nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,349 @@
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    // Transform (pose.x, pose.y) point from global_frame_
+    // to mask_pose point in mask_frame_
+    tf2::Vector3 point(pose.x, pose.y, 0);
+    point = tf2_transform * point;
+    mask_pose.x = point.x();
+    mask_pose.y = point.y();
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_pose = pose;
+  }
+
+  return true;
+}
+
+bool SpeedFilter::worldToMask(double wx, double wy, unsigned int & mx, unsigned int & my) const
+{
+  double origin_x = filter_mask_->info.origin.position.x;
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (wx < origin_x || wy < origin_y) {
+    return false;
+  }
+
+  mx = std::round((wx - origin_x) / resolution);
+  my = std::round((wy - origin_y) / resolution);
+  if (mx >= size_x || my >= size_y) {
+    return false;
+  }
+
+  return true;
+}
+
+inline int8_t SpeedFilter::getMaskData(
+  const unsigned int mx, const unsigned int my) const
+{
+  return filter_mask_->data[my * filter_mask_->info.width + mx];
+}
+
+void SpeedFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      logger_, *(clock_), 2000,
+      "SpeedFilter: Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(pose, mask_pose)) {
+    return;
+  }
+
+  // Converting mask_pose robot position to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  unsigned int mask_robot_i, mask_robot_j;
+  if (!worldToMask(mask_pose.x, mask_pose.y, mask_robot_i, mask_robot_j)) {
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed and
+  // calculating speed limit value
+  int8_t speed_mask_data = getMaskData(mask_robot_i, mask_robot_j);
+  if (speed_mask_data == SPEED_MASK_NO_LIMIT) {
+    // Corresponding filter mask cell is free.
+    // Setting no speed limit there.
+    speed_limit_ = NO_SPEED_LIMIT;
+  } else if (speed_mask_data == SPEED_MASK_UNKNOWN) {
+    // Corresponding filter mask cell is unknown.
+    // Do nothing.
+    RCLCPP_ERROR(
+      logger_,
+      "SpeedFilter: Found unknown cell in filter_mask[%i, %i], "
+      "which is invalid for this kind of filter",
+      mask_robot_i, mask_robot_j);
+    return;
+  } else {
+    // Normal case: speed_mask_data in range of [1..100]
+    speed_limit_ = speed_mask_data * multiplier_ + base_;
```

Comment:
Make sure to document that the base / multiplier values are from 0-100% not 0-1.0 ratio

Reply:
Added a note in parameters description page: https://github.com/ros-planning/navigation.ros.org/pull/122/commits/e178508382727e5ab0cab3f631d8b5960d941259#diff-03647350725051ded3f03120f8c8f447eca3f36c4b5df2f991a022174b3496a3R48